python/ray/_private/log_rotation_with_pipe.py

```diff
@@ -0,0 +1,142 @@
+        self.name = log_name
+        self.write_fd = write_fd
+        self.stream = os.fdopen(write_fd, "w")
+
+    def name(self):
+        return self.name
+
+    def fileno(self):
+        return self.write_fd
+
+    def write(self, s: str) -> int:
+        return self.stream.write(s)
+
+    def flush(self):
+        self.stream.flush()
+
+    def close(self):
+        self.stream.close()
+
+    def isatty(self) -> bool:
+        return False
+
+    def readable(self) -> bool:
+        return False
+
+    def writable(self) -> bool:
+        return True
+
+    def seekable(self) -> bool:
+        return False
+
+
+def open_pipe_with_file_handlers(log_name: str, handlers) -> IO[AnyStr]:
+    """Stream content into pipe, which will be listened and dumped to provided
+    handlers."""
+    read_fd, write_fd = os.pipe()
+
+    log_content = []
+    stopped = threading.Event()
+    cond = threading.Condition()
+
+    # Have to use different logger names for different sinks.
+    logger = logging.getLogger(f"logger-{log_name}")
+    logger.setLevel(logging.INFO)
+    for handler in handlers:
+        logger.addHandler(handler)
+
+    # Setup read thread, which continuous read content out of pipe and append to buffer.
+    def read_log_content_from_pipe():
+        with os.fdopen(read_fd, "r") as pipe_reader:
+            while True:
+                line = pipe_reader.readline()
+
+                with cond:
+                    # An empty line is read over, which indicates write side has
+                    # closed.
+                    if line == _EOF_TOKEN:
+                        stopped.set()
+                        cond.notify()
+                        return
+
+                    # Only buffer new line when not empty.
+                    stripped = line.strip()
+                    if not stripped:
+                        continue
+
+                    # Append content and continue.
+                    log_content.append(stripped)
+                    cond.notify()
+
+    # Setup logging thread, which continues read content out of log buffer and persist
+    # via logger.
+    # Two threads are used here to make sure continuous read from pipe thus avoid
+    # blocking write from application.
+    def dump_log_content_to_buffer():
+        # Current line to log.
+        cur_content = None
+
+        while True:
+            with cond:
+                while not log_content and not stopped.is_set():
+                    cond.wait()
+
+                if log_content:
+                    cur_content = log_content.pop(0)
+                else:
+                    assert stopped.is_set()
+                    return
+
+            # Log and perform IO operation out of critical section.
+            logger.info(cur_content)
+
+    # Python `atexit` hook only invokes when no other non-daemon threads running.
+    read_thread = threading.Thread(target=read_log_content_from_pipe, daemon=True)
```

Comment:
Actually thinking more, I think it's better to implement these in C++ so that the same implementation can be used for all language frontends (e.g. python , java, c++). The current way only works for python workers. I think implementing in C++ should be too much harder?

Reply:
OK, I will create another PR implemented in C++, I will try to get it out before tomorrow.